samples/extauthz/src/main.go

```diff
@@ -0,0 +1,232 @@
+							},
+						},
+					},
+				},
+			},
+			Status: &status.Status{Code: int32(rpc.OK)},
+		}, nil
+	}
+
+	log.Printf("[gRPCv3][denied]: %s", l)
+	return &authv3.CheckResponse{
+		HttpResponse: &authv3.CheckResponse_DeniedResponse{
+			DeniedResponse: &authv3.DeniedHttpResponse{
+				Status: &typev3.HttpStatus{Code: typev3.StatusCode_Forbidden},
+				Body:   denyBody,
+				Headers: []*corev3.HeaderValueOption{
+					{
+						Header: &corev3.HeaderValue{
+							Key:   resultHeader,
+							Value: resultDenied,
+						},
+					},
+				},
+			},
+		},
+		Status: &status.Status{Code: int32(rpc.PERMISSION_DENIED)},
+	}, nil
+}
+
+// ServeHTTP implements the HTTP check request.
+func (s *ExtAuthzServer) ServeHTTP(response http.ResponseWriter, request *http.Request) {
+	l := fmt.Sprintf("%s %s%s, headers: %v\n", request.Method, request.Host, request.URL, request.Header)
+	if allowedValue == request.Header.Get(checkHeader) {
+		log.Printf("[HTTP][allowed]: %s", l)
+		response.Header().Set(resultHeader, resultAllowed)
+		response.WriteHeader(http.StatusOK)
+	} else {
+		log.Printf("[HTTP][denied]: %s", l)
+		response.Header().Set(resultHeader, resultDenied)
+		response.WriteHeader(http.StatusForbidden)
+		response.Write([]byte(denyBody))
+	}
+}
+
+func (s *ExtAuthzServer) startGRPC(address string, wg *sync.WaitGroup) {
+	defer func() {
+		wg.Done()
+		log.Printf("Stopped gRPC server")
+	}()
+
+	listener, err := net.Listen("tcp", address)
+	if err != nil {
+		log.Fatalf("Failed to start gRPC server: %v", err)
+		return
+	}
+	// Store the port for test only.
+	s.grpcPort <- listener.Addr().(*net.TCPAddr).Port
+
+	server := grpc.NewServer()
+	authv2.RegisterAuthorizationServer(server, s.grpcV2)
+	authv3.RegisterAuthorizationServer(server, s.grpcV3)
+
+	log.Printf("Starting gRPC server at %s", listener.Addr())
+	if err := server.Serve(listener); err != nil {
+		log.Fatalf("Failed to serve gRPC server: %v", err)
+		return
+	}
+}
+
+func (s *ExtAuthzServer) startHTTP(address string, wg *sync.WaitGroup) {
+	defer func() {
+		wg.Done()
+		log.Printf("Stopped HTTP server")
+	}()
+
+	listener, err := net.Listen("tcp", address)
+	if err != nil {
+		log.Fatalf("Failed to create HTTP server: %v", err)
+	}
+	// Store the port for test only.
+	s.httpPort <- listener.Addr().(*net.TCPAddr).Port
+
+	log.Printf("Starting HTTP server at %s", listener.Addr())
+	if err := http.Serve(listener, s); err != nil {
+		log.Fatalf("Failed to start HTTP server: %v", err)
+	}
+}
+
+func (s *ExtAuthzServer) run(httpAddr, grpcAddr string) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go s.startHTTP(httpAddr, &wg)
+	go s.startGRPC(grpcAddr, &wg)
+	wg.Wait()
+}
+
+func NewExtAuthzServer() *ExtAuthzServer {
+	return &ExtAuthzServer{
+		grpcV2:   &extAuthzServerV2{},
+		grpcV3:   &extAuthzServerV3{},
+		httpPort: make(chan int, 1),
+		grpcPort: make(chan int, 1),
+	}
+}
+
+func main() {
+	flag.Parse()
+	NewExtAuthzServer().run(fmt.Sprintf(":%s", *httpPort), fmt.Sprintf(":%s", *grpcPort))
```

Comment:
do we want to add SIGTERM handle? I don't really care much since its a sample, but for our tests it will mean cleanup takes 30s unless you add handler or terminationGracePeriod=0s

Reply:
sure, updated to handle SIGTERM.